proposals/0066-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,245 @@
+previously requested feature activations, and end with the activation of those
+with enough support. This process takes one epoch. Therefore, it will take at
+least one more epoch than it currently does from the time a key-holder submits
+a feature for activation via CLI to when it is actually activated.
+
+### Signaling Support for Features
+
+With an on-chain reference point to determine the features queued for
+activation, nodes can signal support for specific features in the queue.
+
+A node signals its support for a feature set by sending a transaction to the
+Feature Gate program, signed by their authorized voter, to store a record of
+their supported feature set.
+This supported feature set would be a bit mask of the list of queued features,
+and it would be stored in a PDA derived from their vote address.
+Nodes simply examine the Feature-Queue list, and mark a 1 for any features they
+have in their feature set. The rest are marked 0.
+
+In the previous example, a node running v1.14.25 would signal the following bit
+mask:
+
+```
+1   0   1   0 
+```
+
+Similarly, a node running v1.16.0 would signal:
+
+```
+1   1   1   1 
+```
+
+Validators send these transactions signaling their support for a queued set of
+features on the first slot of every epoch and on startup after any reboot.
+
+Note: If a feature is revoked before it is moved to the immutable Feature-Queue
+list, it is simply removed from the Feature-Request list and never presented to
+nodes. However, if a feature is revoked after it becomes part of the immutable
+Feature-Queue list, it remains in the list, and nodes still signal their
+support if the feature ID is present in their software. However, the runtime
+won’t activate this feature if its corresponding feature account no longer
+exists on-chain.
+This is how the runtime currently handles features and can be observed here:
+<https://github.com/solana-labs/solana/blob/170478924705c9c62dbeb475c5425b68ba61b375/runtime/src/bank.rs#L8119-L8150>.
+
+### Activating Supported Features
+
+Once the epoch concludes, the runtime uses the validator support signals to
+activate the proper features.
+
+To do this, the runtime walks all of the PDAs derived from the vote accounts,
+examines their bit-masks, and sums the stake supporting each feature-gate to
+determine which feature gates meet the desired threshold of supported stake.
+
+For the initial implementation, the threshold will be hard-coded to 95% of
+stake, but future iterations on the Feature Program could allow feature
+key-holders to set a custom threshold for each feature gate
+(including 0%, ie. –yolo).
+
+Once a list of stake-supported features is generated, the runtime rechecks the
+Feature accounts in case any have been revoked. Then the runtime processes the
+activations using the existing feature logic.
+
+At this point, the runtime clears the existing validator support PDA accounts
+and returns to step 1 to generate the next list of queued features.
```

Comment:
Rather than making the runtime clear all the PDAs (writes!), the PDA data could include the epoch number that the account was last updated in, in addition to the bitmask.

Reply:
Wouldn't this cause two potential issues?

1. You couldn't have static sized PDAs anymore, as per [your other comment](https://github.com/solana-foundation/solana-improvement-documents/pull/72#discussion_r1364778010)
2. You'd have perpetually growing PDAs without garbage collection

Another alternative is to use the epoch number in the PDA itself as a seed, but then we end up growing the number of Feature Gate program-owned accounts.

I think it's probably good to have _some_ form of garbage collection, both to keep the number of accounts owned by Feature Gate down and to return rent lamports.

What about _another_ program that can handle garbage collection? The runtime could just assign all them to the "garbage program" instead of wiping their data (not sure if this is more efficient?) and the program could do it's thing. I'm not totally sure how we'd invoke this program though.